File: Misc/release.mjs
import { readFileSync, writeFileSync } from 'fs';
import { dirname } from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const version = process.argv[ 2 ];
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

if (!/^[0-9]+\.[0-9]+$/.test(version))
{
	console.error('Provide version as first argument. Example: 2.0');
	process.exit(1);
}

process.chdir(__dirname);

const buildPropsPath = '../Directory.Build.props';
const buildProps = readFileSync(buildPropsPath).toString();
const newBuildProps = buildProps.replace(
    /<ProjectBaseVersion>.+<\/ProjectBaseVersion>/,
    `<ProjectBaseVersion>${version}</ProjectBaseVersion>`
);

writeFileSync(buildPropsPath, newBuildProps);

execSync(`git add ${buildPropsPath} && git commit --message="Bump version to ${version}"`, { stdio: 'inherit' });
execSync(`git tag "${version}" --message="${version}" --sign`, { stdio: 'inherit' });

File: Misc/release.cs
using System.Diagnostics;
using System.Text.RegularExpressions;

if (args.Length == 0 || !Regex.IsMatch(args[0], @"^[0-9]+\.[0-9]+$"))
{
    Console.Error.WriteLine("Provide version as first argument. Example: 2.0");
    Environment.Exit(1);
}

var version = args[0];
var buildPropsPath = "Directory.Build.props";

if (!File.Exists(buildPropsPath))
{
    Console.Error.WriteLine("Directory.Build.props not found. Run this script from the repository root.");
    Environment.Exit(1);
}

var buildProps = File.ReadAllText(buildPropsPath);
var newBuildProps = Regex.Replace(
    buildProps,
    @"<ProjectBaseVersion>.+</ProjectBaseVersion>",
    $"<ProjectBaseVersion>{version}</ProjectBaseVersion>"
);

File.WriteAllText(buildPropsPath, newBuildProps);

RunGitCommand($"add {buildPropsPath}");
RunGitCommand($"commit --message \"Bump version to {version}\"");
RunGitCommand($"tag \"{version}\" --message \"{version}\" --sign");

static void RunGitCommand(string arguments)
{
    using var process = new Process
    {
        StartInfo = new ProcessStartInfo
        {
            FileName = "git",
            Arguments = arguments,
            UseShellExecute = false,
            RedirectStandardOutput = false,
            RedirectStandardError = false
        }
    };

    process.Start();
    process.WaitForExit();

    if (process.ExitCode != 0)
    {
        Environment.Exit(process.ExitCode);
    }
}
